nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{U128, U64};
+use near_sdk::collections::{Vector};
+use near_sdk::{env};
+use std::convert::{TryFrom,TryInto};
+
+use std::fmt;
+use crate::constants::*;
+use crate::pool::*;
+use crate::*;
+
+#[derive(Debug)]
+pub enum Mean {
+    M_1MIN,
+    M_5MIN,
+    M_1H,
+    M_12H,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Twap {
+    // To check if array is populated or not
+    // note: it can be used later when we will have variable length storage array/
+    populated: usize,
+    // last updated index in observation array
+    last_updated_index: usize,
+    // bool to check if previous values are overwritten by new one
+    // i.e MAX_LENGTH of array is full and we start storing observation from `0` index
+    pivoted: bool,
+    // observation array
+    pub observations: Vec<Observation>,
+
+    // mean of last 1 minutes of readings
+    pub mean_1min: (U128, U128),
+    // mean of last 5 minutes of readings
+    pub mean_5min: (U128, U128),
+    // mean of last 1 hour of readings
+    pub mean_1h: (U128, U128),
+    // mean of last 12 hours of readings
+    pub mean_12h: (U128, U128)
+}
+
+impl Twap {
+
+    pub fn new() -> Self {
+        Self {
+            populated: 0,
+            last_updated_index: 0,
+            pivoted: false,
+            observations: Vec::new(),
+            mean_1min: (U128(0), U128(0)),
+            mean_5min: (U128(0), U128(0)),
+            mean_1h: (U128(0), U128(0)),
+            mean_12h: (U128(0), U128(0))
+        }
+    }
+
+    /**
+    Initialize the oracle array by writing the first slot.
+    Called once for the lifecycle of the observations array.
+    Parameters:
+    + `time`: The time of the oracle initialization..
+    + `price1`:  price of first token
+    + `price2`: price of second token.
+    */
+    pub fn initialize(
+        &mut self,
+        time: u64,
+        price1: u128,
+        price2: u128
+    ) -> usize
+    {
+        self.observations.push( Observation {
+            block_timestamp: U64(time),
+            num_of_observations: U128(1),
+            price1_cumulative: U128(price1),
+            price2_cumulative: U128(price2),
+        });
+        self.last_updated_index = 0;
+        return 0;
+    }
```

Comment:
Why not moving this code to `new`?

Reply:
because `new()` gives blank vector of observations, and initialize() store first value, when pool is initialized swap is not performed therefore we cannot have it in new().
when the swap is performed we will call log_observation function which will handle initialize() and write()